core/src/main/java/com/linecorp/armeria/server/annotation/ServerSentEventResponseConverterFunction.java

```diff
@@ -0,0 +1,119 @@
+                                        HttpHeaders headers,
+                                        @Nullable Object result,
+                                        HttpHeaders trailingHeaders) throws Exception {
+        final MediaType contentType = headers.contentType();
+        if (contentType != null && contentType.is(MediaType.EVENT_STREAM)) {
+            if (result instanceof Publisher) {
+                return streamingFrom((Publisher<?>) result, headers, trailingHeaders,
+                                     ServerSentEventResponseConverterFunction::toHttpData);
+            }
+            if (result instanceof Stream) {
+                return streamingFrom((Stream<?>) result, headers, trailingHeaders,
+                                     ServerSentEventResponseConverterFunction::toHttpData,
+                                     ctx.blockingTaskExecutor());
+            }
+            return HttpResponse.of(headers, toHttpData(result), trailingHeaders);
+        }
+
+        if (result instanceof ServerSentEvent) {
+            return HttpResponse.of(toMutableHeaders(headers).contentType(MediaType.EVENT_STREAM),
+                                   toHttpData(result), trailingHeaders);
+        }
+
+        return ResponseConverterFunction.fallthrough();
+    }
+
+    private static HttpData toHttpData(@Nullable Object content) {
+        if (content == null) {
+            return HttpData.EMPTY_DATA;
+        }
+
+        final StringBuilder sb = new StringBuilder();
+        if (content instanceof ServerSentEvent) {
+            final ServerSentEvent<?> sse = (ServerSentEvent<?>) content;
+
+            // Write a comment first because a user might want to explain his or her event at first line.
+            sse.comment().ifPresent(comment -> appendField(sb, "", comment, false));
+
+            sse.id().ifPresent(id -> appendField(sb, "id", id, true));
+            sse.event().ifPresent(event -> appendField(sb, "event", event, true));
+            sse.dataAsStringUtf8().ifPresent(data -> appendField(sb, "data", data, true));
+
+            // Reconnection time, in milliseconds.
+            sse.retry().ifPresent(retry -> sb.append("retry:").append(retry.toMillis()).append(LF));
+        } else {
+            appendField(sb, "data", String.valueOf(content), true);
+        }
+
+        final String event = sb.toString();
+        return event.isEmpty() ? HttpData.EMPTY_DATA : HttpData.ofUtf8(event);
+    }
+
+    private static void appendField(StringBuilder sb, String name, String value,
+                                    boolean emitFieldForEmptyValue) {
+        if (value.isEmpty()) {
+            if (emitFieldForEmptyValue) {
+                // Emit name only if the value is an empty string.
+                sb.append(name).append(LF);
+            }
+        } else {
+            sb.append(name)
+              .append(':')
+              .append(lineFeed.matcher(value).replaceAll('\n' + name + ':'))
```

Comment:
Could this be converted into multiple `sb.append()`s without string concatenation? For example, you could use `String.split()`.